### ISSUE TITLE
[OpenMP] Libomptarget depends on `libffi`

### ISSUE THREAD

**Issue by jhuber6** [llvm/llvm-project#88738]
We currently have a dependency on `libffi` in the CPU version of the offloading library. The dependency on `libffi` is unnecessary and also prevents us from having a unified way to pass  arguments to the offloading library. When we emit "kernels" for the CPU they are simply functions in a shared library, see https://godbolt.org/z/5j9r1x9xq.

```llvm
define weak_odr protected void @__omp_offloading_10302_bd4d1_main_l3(ptr %dyn_ptr, i64 %x) {
entry:
  %dyn_ptr.addr = alloca ptr, align 8
  %x.addr = alloca i64, align 8
  store ptr %dyn_ptr, ptr %dyn_ptr.addr, align 8
  store i64 %x, ptr %x.addr, align 8
  %0 = load i32, ptr %x.addr, align 4
  ret void
}
```

What we should do instead is emit this kernel with a known argument that is simply a natively packed struct of all the arguments. This would allow us to pass the arguments as simply a pointer and size, similarly to how both CUDA and HSA allow you to pass arguments.

```llvm
define weak_odr protected void @__omp_offloading_10302_bd4d1_main_l3(ptr noundef %args) {
entry:
  %args.addr = alloca ptr, align 8
  store ptr %args, ptr %args.addr, align 8
  %0 = load ptr, ptr %args.addr, align 8
  %dyn_ptr = getelementptr inbounds %struct.Args, ptr %0, i32 0, i32 0
  %x = getelementptr inbounds %struct.Args, ptr %0, i32 0, i32 8
  ret void
}
```

The code that generates this prologue is at https://github.com/llvm/llvm-project/blob/main/clang/lib/CodeGen/CGStmtOpenMP.cpp#L617, which is shared between all captured OpenMP statements.

**Comment by jhuber6** [comment]
@alexey-bataev I'm not overly familiar with how this should look from the `clang` side. Looking at https://github.com/llvm/llvm-project/blob/main/clang/lib/CodeGen/CGStmtOpenMP.cpp#L617 I'm thinking that all we need to do is adjust the function type and arguments. The function is created here https://github.com/llvm/llvm-project/blob/main/clang/lib/CodeGen/CGStmtOpenMP.cpp#L525, but I don't know what the rules are. Ideally I could just combine all the arguments into a struct and then provide the same interface by getting the arguments from a GEP, but I don't know how to do that given the Clang interface. Some pointers would be appreciated.

**Comment by alexey-bataev** [comment]
> @alexey-bataev I'm not overly familiar with how this should look from the `clang` side. Looking at https://github.com/llvm/llvm-project/blob/main/clang/lib/CodeGen/CGStmtOpenMP.cpp#L617 I'm thinking that all we need to do is adjust the function type and arguments. The function is created here https://github.com/llvm/llvm-project/blob/main/clang/lib/CodeGen/CGStmtOpenMP.cpp#L525, but I don't know what the rules are. Ideally I could just combine all the arguments into a struct and then provide the same interface by getting the arguments from a GEP, but I don't know how to do that given the Clang interface. Some pointers would be appreciated.

You need to build the artificial record and put all the parameters to the member fields. Then need to adjust the type of the function and pass this record argument.  Check https://github.com/llvm/llvm-project/commit/e9c7291cb25f071f1a1dfa4049ed9f5a8a217b3e, looks like it is something you want to do

**Comment by alexey-bataev** [comment]
> @alexey-bataev I looked into that patch and got it working for the basic captured context, however a (semi) recent patch added a forced argument to every target region in
> 
> https://github.com/llvm/llvm-project/blob/35e6bae62c8effa364ded79cc3b0bb988ea88998/clang/lib/Sema/SemaOpenMP.cpp#L4270
> 
> . This now shows up as a separate function argument. The approach you linked relies on using the `RecordDecl` for the `__context` variable in the captured region to handle the arguments. However, because of the `dyn_ptr` there will always be one argument that's not in the struct.
> How would you get around this? I can't make a new `RecordDecl` inside of `CGOpenMPRuntime` because all of the declarations are expected to be constant. I also can't just create a new variable in the scope because this thing needs to be the first argument.

You can try to add it into the `__context` variable as a part of the captured recorddecl

**Comment by alexey-bataev** [comment]
> > You can try to add it into the `__context` variable as a part of the captured recorddecl
> 
> How would that be done? I looked into the surrounding code and couldn't find an easy place to do that. I know we have these `Start` and `End` functions for the surrounding scope, then there's some code that actually captures the things in the block scope. There's something like `createCapture` used elsewhere in the file but I couldn't find how that would be added to the current record decl that's being created. Maybe I'm just missing some accessor that lets us get to the record decl once we call `ActOnCapturedRegionStart`?

Check SemaOpenMP::ActOnOpenMPRegionStart, it adds several fields to the record, like gtid, btid, etc.

**Comment by jhuber6** [comment]
> Check SemaOpenMP::ActOnOpenMPRegionStart, it adds several fields to the record, like gtid, btid, etc.

Code like this https://github.com/llvm/llvm-project/blob/35e6bae62c8effa364ded79cc3b0bb988ea88998/clang/lib/Sema/SemaOpenMP.cpp#L4251?

I think this is the same handling that the `dyn_ptr` argument is using. These are an entirely different `param` to the captured region and aren't put into the `__context` struct. It shows up like this in the AST:
```
CapturedDecl 0x56065cbc56d8 <<invalid sloc>> <invalid sloc> nothrow
|-ImplicitParamDecl 0x56065cbc5740 <line:6:1> col:1 implicit dyn_ptr 'void *const restrict' 
`-ImplicitParamDecl 0x56065cbd9030 <col:1> col:1 implicit __context 'struct (unnamed at input.c:6:1) *const restrict'
```

**Comment by alexey-bataev** [comment]
> > Check SemaOpenMP::ActOnOpenMPRegionStart, it adds several fields to the record, like gtid, btid, etc.
> 
> Code like this
> 
> https://github.com/llvm/llvm-project/blob/35e6bae62c8effa364ded79cc3b0bb988ea88998/clang/lib/Sema/SemaOpenMP.cpp#L4251
> 
> ?
> I think this is the same handling that the `dyn_ptr` argument is using. These are an entirely different `param` to the captured region and aren't put into the `__context` struct. It shows up like this in the AST:
> 
> ```
> CapturedDecl 0x56065cbc56d8 <<invalid sloc>> <invalid sloc> nothrow
> |-ImplicitParamDecl 0x56065cbc5740 <line:6:1> col:1 implicit dyn_ptr 'void *const restrict' 
> `-ImplicitParamDecl 0x56065cbd9030 <col:1> col:1 implicit __context 'struct (unnamed at input.c:6:1) *const restrict'
> ```

Ah, ok, you want to make it part of the __context? Then you need to build an artificial variable in the outer scope and then capture it in the OpenMP context using SemaRef.MarkVariableReferenced. Check SemaOpenMP.cpp, lines 4954-4969, it has something similar I assume

**Comment by jhuber6** [comment]
> Ah, ok, you want to make it part of the __context? Then you need to build an artificial variable in the outer scope and then capture it in the OpenMP context using SemaRef.MarkVariableReferenced. Check SemaOpenMP.cpp, lines 4954-4969, it has something similar I assume

I tried putting this in place of where we currently handle the `dyn_ptr` but it doesn't show up in the `__context` struct. You mentioned "OuterScope" which I interpreted to be creating this before we call `ActOnCapturedRegionStart`. Is there somewhere else I need to stick this so the scope recognizes it as captured?
```c++
auto *VD = buildVarDecl(SemaRef, DSAStack->getConstructLoc(), VoidPtrTy,
                        "dyn_ptr");
auto *RE = buildDeclRefExpr(SemaRef, VD, VD->getType(), VD->getLocation());
SemaRef.MarkVariableReferenced(VD->getLocation(), VD);
```

**Comment by alexey-bataev** [comment]
> > Ah, ok, you want to make it part of the __context? Then you need to build an artificial variable in the outer scope and then capture it in the OpenMP context using SemaRef.MarkVariableReferenced. Check SemaOpenMP.cpp, lines 4954-4969, it has something similar I assume
> 
> I tried putting this in place of where we currently handle the `dyn_ptr` but it doesn't show up in the `__context` struct. You mentioned "OuterScope" which I interpreted to be creating this before we call `ActOnCapturedRegionStart`. Is there somewhere else I need to stick this so the scope recognizes it as captured?
> 
> ```c++
> auto *VD = buildVarDecl(SemaRef, DSAStack->getConstructLoc(), VoidPtrTy,
>                         "dyn_ptr");
> auto *RE = buildDeclRefExpr(SemaRef, VD, VD->getType(), VD->getLocation());
> SemaRef.MarkVariableReferenced(VD->getLocation(), VD);
> ```

It should be created before call of ActOnCapturedRegionStart, but marked as referenced after ActOnCapturedRegionStart (and before ActOnCapturedRegionEnd()).

**Comment by jhuber6** [comment]
> It should be created before call of ActOnCapturedRegionStart, but marked as referenced after ActOnCapturedRegionStart (and before ActOnCapturedRegionEnd()).

Okay, that makes sense. I moved that after the scope beginning and hit an assertion. Guessing that I need to register this in the map somehow?
```
DeclRefExpr for Decl not entered in LocalDeclMap?
UNREACHABLE executed at /home/jhuber/Documents/llvm/llvm-project/clang/lib/CodeGen/CGExpr.cpp:3086!
```

(Thanks for walking me through this, much of `clang` is a mystery to me).

**Comment by alexey-bataev** [comment]
> > It should be created before call of ActOnCapturedRegionStart, but marked as referenced after ActOnCapturedRegionStart (and before ActOnCapturedRegionEnd()).
> 
> Okay, that makes sense. I moved that after the scope beginning and hit an assertion. Guessing that I need to register this in the map somehow?
> 
> ```
> DeclRefExpr for Decl not entered in LocalDeclMap?
> UNREACHABLE executed at /home/jhuber/Documents/llvm/llvm-project/clang/lib/CodeGen/CGExpr.cpp:3086!
> ```
> 
> (Thanks for walking me through this, much of `clang` is a mystery to me).

Yes, you need to create an artificial DeclStmt statement and emit code for it to make this variable defined in the function scope. Or emit it yourself somehow.

**Comment by jhuber6** [comment]
> Yes, you need to create an artificial DeclStmt statement and emit code for it to make this variable defined in the function scope. Or emit it yourself somehow.

I see some code where we create a new `DeclStmt` like here https://github.com/llvm/llvm-project/blob/35e6bae62c8effa364ded79cc3b0bb988ea88998/clang/lib/Sema/SemaOpenMP.cpp#L9804.

Do I simply need to create one for myself with the single variable as the list of declarations?

**Comment by alexey-bataev** [comment]
> > Yes, you need to create an artificial DeclStmt statement and emit code for it to make this variable defined in the function scope. Or emit it yourself somehow.
> 
> I see some code where we create a new `DeclStmt` like here
> 
> https://github.com/llvm/llvm-project/blob/35e6bae62c8effa364ded79cc3b0bb988ea88998/clang/lib/Sema/SemaOpenMP.cpp#L9804
> 
> .
> Do I simply need to create one for myself with the single variable as the list of declarations?

Yes. But you need to store it somewhere and call a codegen for it

**Comment by jhuber6** [comment]
> Yes. But you need to store it somewhere and call a codegen for it

Seems the other places store these in the clauses they create. Is it not possible to add this to the existing captured statement that is used in stuff like `ActOnOpenMPTargetDirective`? Also what does it mean to call codegen for it.

**Comment by jhuber6** [comment]
Okay, I realized what the issue is.

We normally promote all arguments it `uintptr_t` when calling outlined functions. The numbers I observed were actually 32-bit values with 32 bits of random stack garbage because I was inspecting it as a 64-bit value. The kernel call will use i64 spacing and then demote to the lower type when it's used. The `RecordDecl` here is instead capturing everything as its native type, i32, which means that we are getting the arguments out of order. I can either hack around this in the runtime or somehow figure out how to do it in Clang. Given that I don't think we can even create a new RecordDecl at the moment we want to emit this function prologue during CodeGen I might just do the latter.

**Comment by alexey-bataev** [comment]
> Okay, I realized what the issue is.
> 
> 
> 
> We normally promote all arguments it `uintptr_t` when calling outlined functions. The numbers I observed were actually 32-bit values with 32 bits of random stack garbage because I was inspecting it as a 64-bit value. The kernel call will use i64 spacing and then demote to the lower type when it's used. The `RecordDecl` here is instead capturing everything as its native type, i32, which means that we are getting the arguments out of order. I can either hack around this in the runtime or somehow figure out how to do it in Clang. Given that I don't think we can even create a new RecordDecl at the moment we want to emit this function prologue during CodeGen I might just do the latter.

Yes, expected something like that. We can create new RecordDecl, the fields should be packed as ints of pointers width and copy the data there. But it may consume extra time for field-by-field copying.

**Comment by jhuber6** [comment]
> Yes, expected something like that. We can create new RecordDecl, the fields should be packed as ints of pointers width and copy the data there. But it may consume extra time for field-by-field copying.

Could you point me to where we do something like this? I tried to look at where other places created a RecordDecl but they all relied on some kind of `DeclContext` that was expected to be `const` by the time we get to codegen.
